File: Python/companies_server.py
# This companies_server.py file utilizes the generated .py files by grpc tools to run a server
# The grpc tools generate companies_pb2.py and companies_pb2_grpc.py according to companies.proto
# companies_pb2.py contains classes for the messages defined in companies.proto
# companies_pb2_grpc.py contains client stub and servicer classes for the services defined in companies.proto
# A function from companies_resources.py is also called in this file to load some data
# This file is part of a grpc demo
# In order to run this program, some grpc modules in python should be firstly installed,
# please refer to the grpc tutorials on http://www.grpc.io/docs/quickstart/python.html


"""The Python implementation of the gRPC companies information server."""

from concurrent import futures
import time

import grpc

import companies_pb2
import companies_pb2_grpc
import companies_resources

_ONE_DAY_IN_SECONDS = 60 * 60 * 24


def get_founding_year(companies_db, input_name):
  """Returns the founding year of the given company name."""
  for company in companies_db:
    if company.name.name == input_name.name:
      return company.foundingYear
  #for simplicity, a default year 2900 is returned if no matched company is found
  return companies_pb2.FoundingYear(year=2900)


class CompaniesServicer(companies_pb2_grpc.CompaniesServicer):
  """Provides methods that implement functionality of companies information server."""

  def __init__(self):
    self.db = companies_resources.read_companies_database()

  def GetFoundingYear(self, request, context):
    founding_year = get_founding_year(self.db, request)
    return founding_year

  def ListCompanies(self, request, context):
    for company in self.db:
      if company.foundingYear == request:
        yield company.name

  def CalcAverageAge(self, request_iterator, context):
    current_year = 2017
    total_age = 0
    company_count = 0

    for company_name in request_iterator:
      for company in self.db:
        if company.name.name == company_name.name:
          total_age += (current_year - company.foundingYear.year)
          company_count += 1
          break

    average_age = int(total_age / company_count)
    return companies_pb2.AverageAge(age=average_age)  

  def GetCompanyInformation(self, request_iterator, context):
    for company_name in request_iterator:
      found = False
      for company in self.db:
        if company.name == company_name:
          yield company
          found = True
          break
      if not found:
        yield companies_pb2.Company(
            name = companies_pb2.Name(name="not found"),
            foundingYear = companies_pb2.FoundingYear(year=2900),
            marketValue = 0)


def serve():
  server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
  companies_pb2_grpc.add_CompaniesServicer_to_server(
      CompaniesServicer(), server)
  server.add_insecure_port('[::]:50051')
  server.start()
  print("Server has started, listenning in port 50051")
  try:
    while True:
      time.sleep(_ONE_DAY_IN_SECONDS)
  except KeyboardInterrupt:
    server.stop(0)
    print("Server has stopped")

if __name__ == '__main__':
  serve()

File: Java/src/main/java/io/grpc/grpcDemo1/CompaniesServer.java
//This CompaniesServer.java file is part of a grpc demo
//It utilizes the generated source files by grpc tools to run a server
//The generated source files are in the folder of target/generated-sources/protobuf
//The grpc tools generate some classes according to the messages defined in companies.proto
//These classes definitions are in the folder of target/generated-sources/protobuf/java
//The grpc tools also generate a classes according to the service defined in companies.proto
//This class definition is in the folder of target/generated-sources/protobuf/grpc-java
//In order to run the program, some grpc related libraries should be included in the project
//Therefore this project was set up as a Maven project and its pom.xml file includes suitable depencencies
//For further information, please refer to the grpc tutorials on http://www.grpc.io

package io.grpc.grpcDemo1;

import io.grpc.Server;
import io.grpc.ServerBuilder;
import io.grpc.stub.StreamObserver;
import java.io.IOException;
import java.net.URL;
import java.util.Collection;
//import java.util.List;
//import java.util.logging.Level;
import java.util.logging.Logger;

public class CompaniesServer {
	
	private static final Logger logger = Logger.getLogger(CompaniesServer.class.getName());

	private final int port;
	private final Server server;
	private final CompanyService service;
	
	public CompaniesServer(int port) throws IOException {
	    this(port, CompaniesUtil.getDefaultCompaniesFile());
	}
	
	/** Create a company information server listening on {@code port} using {@code companiesFile} database. */
	public CompaniesServer(int port, URL companiesFile) throws IOException {
	  this(ServerBuilder.forPort(port), port, CompaniesUtil.parseCompanies(companiesFile));
	}	
	
	/** Create a company information server using serverBuilder as a base and companies as data. */
	public CompaniesServer(ServerBuilder<?> serverBuilder, int port, Collection<Company> companies) {
	  this.port = port;
	  this.service = new CompanyService(companies);
	  server = serverBuilder.addService(service).build();
	}
	
	/** Start serving requests. */
	public void start() throws IOException{
		server.start();
		System.out.println("Server started, listening on " + port);
//		logger.info("Server started, listening on " + port);
		Runtime.getRuntime().addShutdownHook(new Thread() {
			@Override
		    public void run() {
		      // Use stderr here since the logger may has been reset by its JVM shutdown hook.
		      System.err.println("*** shutting down gRPC server since JVM is shutting down");
		      CompaniesServer.this.stop();
		      System.err.println("*** server shut down");
		    }
		});
	}
	
	/** Stop serving requests and shutdown resources. */
	public void stop() {
	    if (server != null) {
	        server.shutdown();
	    }
	}
	
	/**
	 * Await termination on the main thread since the grpc library uses daemon threads.
	 */
	private void blockUntilShutdown() throws InterruptedException {
	    if (server != null) {
	        server.awaitTermination();
	    }
	}
	
//	for debugging, print all the company data in the json database
	private void printData() {
		System.out.println(server.getServices().get(0));
		for(Company company : service.companies) {
			System.out.println("Company " + company.getName().getName() + 
					", founded in year " + company.getFoundingYear().getYear() + 
					", market value is " + company.getMarketValue());
		}
		System.out.println();
	}
	
	public static void main(String[] args) throws Exception{
		
		//		System.out.println("test");
		CompaniesServer server = new CompaniesServer(50051);
		server.printData();		//check if the data is ready and correct
		server.start();
		server.blockUntilShutdown();
	}
	
	private static class CompanyService extends CompaniesGrpc.CompaniesImplBase {
		private final Collection<Company> companies;
		
		CompanyService(Collection<Company> companies) {
		  this.companies = companies;
		}
		
		@Override
		public void getFoundingYear(Name request, StreamObserver<FoundingYear> responseObserver) {
			responseObserver.onNext(checkFoundingYear(request));
			responseObserver.onCompleted();
		}
		
		@Override
		public void listCompanies(FoundingYear request, StreamObserver<Name> responseObserver) {
			for(Company company : companies) {
				if(company.getFoundingYear().getYear() == request.getYear()) {
					responseObserver.onNext(company.getName());
				}
			}
			responseObserver.onCompleted();
		}
		
		@Override
		public StreamObserver<Name> calcAverageAge(final StreamObserver<AverageAge> responseObserver) {
			return new StreamObserver<Name>() {
				public int totalAge = 0;
				int count = 0;
				
//				@Override
		        public void onNext(Name name) {
		        	int age = 2017 - checkFoundingYear(name).getYear();
		        	System.out.println("age is " + age);
		        	if(age != 2017) {
		        		count++;
		        		totalAge += age;
		        	}
		        	System.out.println("Test1: total age is " + this.totalAge + ", count is " + this.count);
		        	System.out.println();
				}
		        
//		        @Override
		        public void onError(Throwable t) {
		        	System.out.println("Calculating average age cancelled");
				}
		        
//		        @Override
		        public void onCompleted() {
		        	System.out.println("total age is " + this.totalAge + ", count is " + this.count);
//		        	AverageAge anAge = AverageAge.newBuilder().setAge(1000).build();
//		        	System.out.println(responseObserver.toString());
//		        	responseObserver.onNext(anAge);
		        	if(count != 0) {
		        		int average = (int)(totalAge / count);
		        		System.out.println("average age " + average);
		        		responseObserver.onNext(AverageAge.newBuilder().setAge(average).build());
		        		responseObserver.onCompleted();
		        	}
		        	else {
		        		responseObserver.onNext(AverageAge.newBuilder().setAge(0).build());
		        		responseObserver.onCompleted();
		        	}
		        }
			}; 
		}
		
		@Override
	    public StreamObserver<Name> getCompanyInformation(final StreamObserver<Company> responseObserver) {
			return new StreamObserver<Name>() {
//				@Override
		        public void onNext(Name name) {
		        	boolean found = false;
		        	for(Company company : companies) {
		        		if(company.getName().getName().equals(name.getName())) {
		        			responseObserver.onNext(company);
		        			found = true;
		        			break;
		        		}
		        	}
		        	if(!found) {
		        		responseObserver.onNext(Company.newBuilder().setName(Name.newBuilder().setName("Not found").build())
		        				.setFoundingYear(FoundingYear.newBuilder().setYear(0).build())
		        				.setMarketValue(0).build());
		        	}
				}
		        
//		        @Override
		        public void onError(Throwable t) {
		        	System.out.println("Getting companies information cancelled");;
				}
		        
//		        @Override
		        public void onCompleted() {
		        	responseObserver.onCompleted();
		        }
			}; 
		}
		
		private FoundingYear checkFoundingYear(Name name) {
//			logger.info("input name " + name.getName());
			for(Company company : companies) {
//				logger.info("company name " + company.getName().getName());
				if(company.getName().getName().equals(name.getName())) {
//					logger.info("company name " + company.getName().getName());
					return company.getFoundingYear();
				}
			}
			
			FoundingYear returnYear = FoundingYear.newBuilder().setYear(0).build();
//			logger.info("Not found year " + returnYear.getYear());
			return returnYear;
		}
	}
}
